cmd/controller/app/controller.go

```diff
@@ -63,89 +66,145 @@ const controllerAgentName = "cert-manager"
 
-	var wg sync.WaitGroup
-	run := func(_ context.Context) {
-		for n, fn := range controller.Known() {
-			log := log.WithValues("controller", n)
+	g.Go(func() error {
+		<-rootCtx.Done()
+		// allow a timeout for graceful shutdown
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		defer cancel()
 
-			// only run a controller if it's been enabled
-			if !enabledControllers.Has(n) {
-				log.V(logf.InfoLevel).Info("not starting controller as it's disabled")
-				continue
+		if err := server.Shutdown(ctx); err != nil {
+			return err
+		}
+		return nil
+	})
+	g.Go(func() error {
+		log.WithValues("address", ln.Addr()).V(logf.InfoLevel).Info("listening for connections on")
+		if err := server.Serve(ln); err != http.ErrServerClosed {
+			return err
+		}
+		return nil
+	})
+
+	elected := make(chan struct{})
+	if opts.LeaderElect {
+		g.Go(func() error {
+			log.V(logf.InfoLevel).Info("starting leader election")
+			leaderElectionClient, err := kubernetes.NewForConfig(rest.AddUserAgent(kubeCfg, "leader-election"))
+			if err != nil {
+				log.Error(err, "error creating leader election client")
+				return err
 			}
 
-			// don't run clusterissuers controller if scoped to a single namespace
-			if ctx.Namespace != "" && n == clusterissuers.ControllerName {
-				log.V(logf.InfoLevel).Info("not starting controller as cert-manager has been scoped to a single namespace")
-				continue
+			errorCh := make(chan error, 1)
+			if err := startLeaderElection(rootCtx, opts, leaderElectionClient, ctx.Recorder, leaderelection.LeaderCallbacks{
+				OnStartedLeading: func(_ context.Context) {
+					close(elected)
+				},
+				OnStoppedLeading: func() {
+					select {
+					case <-rootCtx.Done():
+						// context was canceled, just return
+						return
+					default:
+						log.V(logf.ErrorLevel).Info("leader election lost")
+						errorCh <- errors.New("leader election lost")
+					}
+				},
+			}); err != nil {
+				return err
 			}
 
-			wg.Add(1)
-			iface, err := fn(ctx)
-			if err != nil {
-				log.Error(err, "error starting controller")
-				os.Exit(1)
+			select {
+			case err := <-errorCh:
+				return err
+			default:
+				return nil
 			}
-			go func(n string, fn controller.Interface) {
-				defer wg.Done()
-				log.V(logf.InfoLevel).Info("starting controller")
-
-				workers := 5
-				err := fn.Run(workers, stopCh)
-
-				if err != nil {
-					log.Error(err, "error starting controller")
-					os.Exit(1)
-				}
-			}(n, iface)
-		}
+		})
+	} else {
+		close(elected)
+	}
 
-		log.V(logf.DebugLevel).Info("starting shared informer factories")
-		ctx.SharedInformerFactory.Start(stopCh)
-		ctx.KubeSharedInformerFactory.Start(stopCh)
-		ctx.GWShared.Start(stopCh)
-		wg.Wait()
-		log.V(logf.InfoLevel).Info("control loops exited")
-		ctx.Metrics.Shutdown(metricsServer)
-		os.Exit(0)
+	select {
+	case <-rootCtx.Done(): // Exit early if the Elected channel gets closed because we are shutting down.
```

Comment:
https://pkg.go.dev/golang.org/x/sync/errgroup#Group.Wait
"Wait blocks until all function calls from the Go method have returned, then returns the first non-nil error (if any) from them."
The errgroup will wait for all goroutines to return, not just those that returned an error. But in case an error is returned, the linked context will be canceled. If the context is canceled (either because of an error or a interrupt signal), we want all goroutines to return -> this is something we have to check for ourselves in our goroutine's implementation by passing the errgroup context & adding them to all select cases.

Reply:
Yeah, but I was thinking of the case where one of the goroutines exits early with `nil`.....for some reason. 
Perhaps if I examined the code, I'd see that that can't ever happen, but nevertheless, the expectation is that all the goroutines are running, otherwise cert-manager will not be functioning properly.

So *ideally*, we'd use something *like* errgroup, but where the exit of one goroutine (with error or nil) would immediatly cancel the group context so that the other goroutines can also exit cleanly.

If you agree with my logic, I don't propose that it should be done in this PR, but perhaps it should be considered for a future PR.